build-data-capturing-gradle-samples/capture-thermal-throttling/gradle-thermal-throttling.gradle.kts

```diff
@@ -0,0 +1,142 @@
+import org.gradle.api.internal.project.ProjectInternal
+import org.gradle.api.services.BuildService
+import org.gradle.api.services.BuildServiceParameters
+import com.gradle.enterprise.gradleplugin.GradleEnterpriseExtension
+import org.gradle.internal.os.OperatingSystem
+import com.gradle.scan.plugin.BuildScanExtension
+
+import java.nio.charset.Charset
+import java.util.concurrent.ConcurrentLinkedQueue
+import java.util.concurrent.Executors
+import java.util.concurrent.ExecutorService
+import java.util.concurrent.TimeUnit
+import java.util.function.Function
+import java.util.Queue
+
+import org.gradle.tooling.events.OperationCompletionListener
+import org.gradle.tooling.events.FinishEvent
+
+/**
+ * This Gradle script captures the thermal throttling level and adds it as a tag.
+ * Some parameters can be tweaked according to your build:
+ * - SAMPLING_INTERVAL_IN_SECONDS, frequency at which the capture command is run
+ * - THROTLLING_LEVEL, list of throttling levels and value ranges (to be compared with captured throttling average value)
+ *
+ * WARNINGS:
+ * - This is supported on MacOS only.
+ */
+
+project.extensions.configure<GradleEnterpriseExtension>() {
+    buildScan {
+        val api = buildScan
+        if (OperatingSystem.current().isMacOsX()) {
+            // Register Thermal throttling service
+            val throttlingServiceProvider =
+                gradle.sharedServices.registerIfAbsent("thermalThrottling", ThermalThrottlingService::class.java) {}
+            // start service
+            (project as ProjectInternal).services.get(BuildEventsListenerRegistry::class.java).onTaskCompletion(throttlingServiceProvider)
+
+            buildScan.buildFinished {
```

Comment:
why do we need to redeclare `buildScan` here? is this because otherwise it will call `gradle.buildFinished`?

Reply:
ah it's a left over from copy pasting Groovy code